### PR TITLE
GH-33032: [C#] Support fixed-size lists

### LINKED ISSUE #32886: [C#] Exception when reading binary arrow file (Value cannot be null. (Parameter 'name'))
Hi everyone,

I'm trying to read binary file using csharp apache arrow library v9.0.0 and I'm facing this exception
```java

Unhandled exception. System.AggregateException: One or more errors occurred. (Value cannot be null. (Parameter 'name'))
 ---> System.ArgumentNullException: Value cannot be null. (Parameter 'name')
   at Apache.Arrow.Field..ctor(String name, IArrowType dataType, Boolean nullable)
   at Apache.Arrow.Ipc.MessageSerializer.FieldFromFlatbuffer(Field flatbufField, DictionaryMemo& dictionaryMemo)
   at Apache.Arrow.Ipc.MessageSerializer.FieldFromFlatbuffer(Field flatbufField, DictionaryMemo& dictionaryMemo)
   at Apache.Arrow.Ipc.MessageSerializer.GetSchema(Schema schema, DictionaryMemo& dictionaryMemo)
   at Apache.Arrow.Ipc.ArrowStreamReaderImplementation.<ReadSchemaAsync>b__10_0(Memory`1 buff)
   at Apache.Arrow.ArrayPoolExtensions.RentReturnAsync(ArrayPool`1 pool, Int32 length, Func`2 action)
   at Apache.Arrow.Ipc.ArrowStreamReaderImplementation.ReadSchemaAsync()
   at Apache.Arrow.Ipc.ArrowStreamReaderImplementation.ReadRecordBatchAsync(CancellationToken cancellationToken)
   at Apache.Arrow.Ipc.ArrowStreamReaderImplementation.ReadNextRecordBatchAsync(CancellationToken cancellationToken) 
```
As far as I do understand, the library is complaining about field name being null, not sure if it's the case since I tried to read the same file using apache arrow golang library and it seems to work without issue.

Please find attached the `sample.arrow` file

Below a sample code I'm using to read this arrow file:
- Csharp sample (fails)
  ```java
  
  using System;
  using System.IO;
  using System.Threading.Tasks;
  using Apache.Arrow;
  using Apache.Arrow.Ipc;
  namespace arrow_csharp_issue
  {
      class Program
      {
          static async Task AsyncMain()
          {
              byte[] bytes = File.ReadAllBytes("./inputs/sample.arrow");
              using (var memoryStream = new MemoryStream(bytes))
              using (var reader = new ArrowStreamReader(memoryStream))     
              { 
               RecordBatch record = await reader.ReadNextRecordBatchAsync();
               Console.WriteLine(record);
              }
          }
          static void Main(string[] args)   
          {
             AsyncMain().Wait();
          }
      }
  } 
  ```
  
- Golang sample (works)
  ```java
  
  package main
  
  import (
  "bytes"
  "fmt"
  "os"
  "github.com/apache/arrow/go/v9/arrow/ipc"
  )
  
  func main() {
    data, err := os.ReadFile("./inputs/sample.arrow")
    if err != nil {
      panic(err)
    }
  
    reader, err := ipc.NewReader(bytes.NewReader(data))
    if err != nil { panic(err) }
    defer reader.Release()
    reader.Next()
    record := reader.Record()
    fmt.Println(record)
  }
  ```
  
   
  
  Thank you

**Reporter**: [helmi](https://issues.apache.org/jira/browse/ARROW-17644)
#### Original Issue Attachments:
- [sample.arrow](https://issues.apache.org/jira/secure/attachment/13049057/sample.arrow)

<sub>**Note**: *This issue was originally created as [ARROW-17644](https://issues.apache.org/jira/browse/ARROW-17644). Please see the [migration documentation](https://github.com/apache/arrow/issues/14542) for further details.*</sub>

### LINKED ISSUE #33032: [C#] FixedSizeList implementation is missing
Hi,

I'm working toward integrating apache arrow c# and I find out that FixedSizeList is not implemented. Is there a plan to implement the missing type ? Otherwise, what are my options to overcome this issue ?

https://issues.apache.org/jira/browse/ARROW-17644

**Reporter**: [helmi](https://issues.apache.org/jira/browse/ARROW-17808)

<sub>**Note**: *This issue was originally created as [ARROW-17808](https://issues.apache.org/jira/browse/ARROW-17808). Please see the [migration documentation](https://github.com/apache/arrow/issues/14542) for further details.*</sub>

### PR DESCRIPTION
**Pull request by CurtHagenlocher**
### What changes are included in this PR?

Support fixed-size lists in the C# implementation.
Adds Archery support for Lists, Structs and Fixed-size Lists in the C# implementation.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

Fixed-size lists are now supported for C#.

* Closes: #33032
* Closes: #32886

### CODE REVIEW

File: csharp/src/Apache.Arrow/Field.cs [hunk 1]
```diff
@@ -35,24 +35,24 @@ public partial class Field
...
+            IReadOnlyDictionary<string, string> metadata, bool copyCollections, bool allowBlankName)
+            : this(name, dataType, nullable, allowBlankName)
         {
             Debug.Assert(copyCollections == false, "This internal constructor is to not copy the collections.");
 
             Metadata = metadata;
         }
 
-        private Field(string name, IArrowType dataType, bool nullable)
+        private Field(string name, IArrowType dataType, bool nullable, bool allowBlankName)
```
Comment (CurtHagenlocher) [inline csharp/src/Apache.Arrow/Field.cs:53]: I concede that this change is a bit odd; the intent was to address #32886.

Comment (CurtHagenlocher) [reply csharp/src/Apache.Arrow/Field.cs:53]: Though perhaps this should be done more broadly as per #36588.

File: csharp/src/Apache.Arrow/Types/FixedSizeListType.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    {
+        public override ArrowTypeId TypeId => ArrowTypeId.FixedSizeList;
+        public override string Name => "fixed_size_list";
+        public int ListSize { get; }
+
+        public Field ValueField => Fields[0];
+
+        public IArrowType ValueDataType => Fields[0].DataType;
+
+        public FixedSizeListType(Field valueField, int listSize)
```
Comment (lidavidm) [inline csharp/src/Apache.Arrow/Types/FixedSizeListType.cs:28]: Worth validating that listSize > 0?

**Review by lidavidm — APPROVED** [review]

File: csharp/src/Apache.Arrow/Types/FixedSizeListType.cs (hunk 2 above)
Comment (CurtHagenlocher) [reply csharp/src/Apache.Arrow/Types/FixedSizeListType.cs:30]: Added. (Also updated with latest source and fixed merge conflict.)

Comment (CurtHagenlocher) [reply csharp/src/Apache.Arrow/Types/FixedSizeListType.cs:30]: Derp.... not sure what I did :(

Comment (CurtHagenlocher) [reply csharp/src/Apache.Arrow/Types/FixedSizeListType.cs:30]: ...ah, it was a conflict with a different change of mine that's been checked in. Fixed.

**Review by lidavidm — APPROVED** [review]
Thanks!

**Comment by lidavidm** [comment]
Alright, looks like integration and docs failures are unrelated

**Pull request merged**